core/src/connection/manager/task.rs

```diff
@@ -218,112 +229,114 @@ where
-                            this.state = State::EstablishedReady { connection: None, event }
+                            this.state = State::Terminating(event)
                         }
                     }
                 }
 
-                State::EstablishedPending(mut connection) => {
-                    // Start by handling commands received from the manager, if any.
+                State::Established { mut connection, event } => {
+                    // Check for commands from the `Manager`.
                     loop {
-                        match Stream::poll_next(Pin::new(&mut this.commands), cx) {
+                        match this.commands.poll_next_unpin(cx) {
                             Poll::Pending => break,
                             Poll::Ready(Some(Command::NotifyHandler(event))) =>
                                 connection.inject_event(event),
-                            Poll::Ready(None) => {
-                                // The manager has dropped the task, thus initiate a
-                                // graceful shutdown of the connection.
+                            Poll::Ready(Some(Command::Close)) => {
+                                // Discard the event, if any, and start a graceful close.
                                 this.state = State::Closing(connection.close());
                                 continue 'poll
                             }
+                            Poll::Ready(None) => {
+                                // The manager has dropped the task; abort.
+                                return Poll::Ready(())
+                            }
                         }
                     }
-                    // Poll the connection for new events.
-                    loop {
+
+                    if let Some(event) = event {
+                        // Send the event to the manager.
+                        match this.events.poll_ready(cx) {
+                            Poll::Pending => {
+                                this.state = State::Established { connection, event: Some(event) };
+                                return Poll::Pending
+                            }
+                            Poll::Ready(result) => {
+                                if result.is_ok() {
+                                    if let Ok(()) = this.events.start_send(event) {
+                                        this.state = State::Established { connection, event: None };
+                                        continue 'poll
+                                    }
+                                }
+                                // The manager is no longer reachable, maybe due to
+                                // application shutdown. Try a graceful shutdown of the
+                                // connection before terminating the task.
+                                this.state = State::Closing(connection.close());
```

Comment:
I know we did it like this so far but I wonder if child resource should linger around and perform I/O when their managing parent has been dropped?

Reply:
I wondered about this as well and am in favour of no longer doing it, because with this PR there is now actually a way to do a "clean" shutdown if you want to by explicitly closing the connections and waiting for the corresponding events.